src/tensor/cache.rs

```diff
@@ -29,80 +75,215 @@ pub(crate) struct AllocationKey {
 /// allocator assumes memory is allocated & deallocated with the same layout.
 /// The value is a list of allocations of that size.
 ///
-/// The prescense of a key in the map, indicates that there is *at least one*
+/// The presence of a key in the map, indicates that there is *at least one*
 /// valid allocation. When the last value is removed from the list, the key
 /// is removed.
+///
+/// Constraint: for a given value of AllocationKey, the following must hold for each key value in
+/// `allocations`:
+///
+/// (instances of key in drop_queue) = allocations[key].ignore_drops + allocations[key].allocations.len()
+#[derive(Debug)]
+pub(crate) struct TensorCache<Ptr: CacheStorage> {
+    allocations: RwLock<BTreeMap<AllocationKey, AllocationGroup<Ptr>>>,
+    enabled: RwLock<bool>,
+
+    drop_queue: RwLock<VecDeque<AllocationKey>>,
+    size: RwLock<usize>,
+    max_size: RwLock<usize>,
+}
+
+pub(crate) trait CacheStorage: Sized {
+    type Output<T>: CacheStorage;
+
+    /// returns the allocations's size in bytes
+    fn size(&self) -> usize;
+
+    /// Unsafely converts the elements of a contiguous collection type to another type. Note:
+    /// **This function is wildly unsafe**, see implementations for details
+    unsafe fn transmute_elements<T>(self) -> Self::Output<T>;
+
+    /// Uses transmute_elements to convert to an element type with alignment `align` before dropping.
+    /// This **must** be a memory safe way to drop self, given the correct alignment
+    unsafe fn drop_with_alignment(self, align: usize) {
```

Comment:
Thoughts on removing default impl of this? This implementation is the CPU one, and I'm not sure we should be doing all of this for cuda.

Reply:
I think that this is correct for cuda, and while it may not be necessary, it doesn't have a significant performance impact as transmute_elements is very fast.